indexing-service/src/main/java/io/druid/indexing/overlord/http/OverlordResource.java

```diff
@@ -575,58 +542,231 @@ public Response getCompleteTasks(
       @Context final HttpServletRequest req
   )
   {
-    final Function<String, Task> taskFunction = taskId -> {
-      final Optional<Task> optionalTask = taskStorageQueryAdapter.getTask(taskId);
-      if (!optionalTask.isPresent()) {
+    return getTasks("complete", null, null, maxTaskStatuses, null, req);
+  }
+
+  @GET
+  @Path("/tasks")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getTasks(
+      @QueryParam("state") final String state,
+      @QueryParam("datasource") final String dataSource,
+      @PathParam("interval") final String interval,
+      @QueryParam("max") final Integer maxCompletedTasks,
+      @QueryParam("type") final String type,
+      @Context final HttpServletRequest req
+  )
+  {
+
+    //check for valid state
+    if (state != null) {
+      List validStates = ImmutableList.of("pending", "waiting", "running", "complete");
+      if (!validStates.contains(StringUtils.toLowerCase(state))) {
+        return Response.status(Status.BAD_REQUEST)
+                       .entity("Bad state query param passed : " + state).entity(ImmutableList.of()).build();
+      }
+    }
+    // early authorization check if datasource != null
+    // fail fast if user not authorized to access datasource
+    if (dataSource != null) {
+      final ResourceAction resourceAction = new ResourceAction(
+          new Resource(dataSource, ResourceType.DATASOURCE),
+          Action.READ
+      );
+      Access authResult = AuthorizationUtils.authorizeResourceAction(
```

Comment:
It's a little weird to mix `final` and non-`final` variables here. Either one is ok (we don't have a strong style guide for whether variables like this should be final or not) but you should at least be consistent within a method.

If it was me writing it then I would do `final` for both :)

Reply:
yes `final` it is.